coordinator/src/position/models.rs

```diff
@@ -49,4 +56,198 @@ pub struct Position {
+}
+
+impl Position {
+    /// Calculates the profit and loss for the coordinator
+    ///
+    /// The position stored represents the values of the trader.
+    pub fn calculate_coordinator_pnl(&self, quote: Quote) -> Result<i64> {
+        let closing_price = match self.closing_price {
+            None => closing_price(self.direction, quote),
+            Some(closing_price) => {
+                Decimal::try_from(closing_price).expect("f32 closing price to fit into decimal")
+            }
+        };
+
+        let average_entry_price = Decimal::try_from(self.average_entry_price)
+            .context("Failed to convert average entry price to Decimal")?;
+
+        let (long_leverage, short_leverage) = match self.direction {
+            Direction::Long => (self.leverage, COORDINATOR_LEVERAGE),
+            Direction::Short => (COORDINATOR_LEVERAGE, self.leverage),
+        };
+
+        // the position in the database is the trader's position, our direction is opposite
+        let direction = self.direction.opposite();
+
+        let pnl = calculate_pnl(
+            average_entry_price,
+            closing_price,
+            self.quantity,
+            long_leverage,
+            short_leverage,
+            direction,
+        )
+        .context("Failed to calculate pnl for position")?;
+
+        Ok(pnl)
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+    use std::str::FromStr;
+
+    #[test]
+    fn given_trader_long_position_when_no_bid_price_change_then_zero_pnl() {
+        let position = Position::dummy()
+            .with_leverage(2.0)
+            .with_quantity(1.0)
+            .with_average_entry_price(1000.0)
+            .with_direction(Direction::Long);
+
+        let quote = dummy_quote(1000, 0);
+
+        let coordinator_pnl = position.calculate_coordinator_pnl(quote).unwrap();
+
+        assert_eq!(coordinator_pnl, 0);
+    }
+
+    #[test]
+    fn given_trader_short_position_when_no_ask_price_change_then_zero_pnl() {
+        let position = Position::dummy()
+            .with_leverage(2.0)
+            .with_quantity(1.0)
+            .with_average_entry_price(1000.0)
+            .with_direction(Direction::Short);
+
+        let quote = dummy_quote(0, 1000);
+
+        let coordinator_pnl = position.calculate_coordinator_pnl(quote).unwrap();
+
+        assert_eq!(coordinator_pnl, 0);
+    }
+
+    /// See also: `given_long_position_when_price_10_pc_up_then_18pc_profit` test in `trade::cfd`
+    #[test]
+    fn given_trader_long_position_when_bid_price_10pc_up_then_coordinator_18pc_loss() {
+        let position = Position::dummy()
+            .with_leverage(2.0)
+            .with_quantity(20000.0)
+            .with_average_entry_price(20000.0)
+            .with_direction(Direction::Long);
+
+        let quote = dummy_quote(22000, 0);
+
+        let coordinator_pnl = position.calculate_coordinator_pnl(quote).unwrap();
+
+        assert_eq!(coordinator_pnl, -9_090_909);
```

Comment:
It's hard for me to follow these amounts :(

Reply:
Ok, it took me a while to understand what is happening and why this value is correct: 

In my words expressed: In this example, the person who went long, bought $20,000 worth of BTC at the price of 20,000, i.e. `1` BTC
At the price of $22,000 he sells $20,000 worth of BTC, i.e. he sells it for `0.909090909` BTC. 
The difference is his profit, i.e. 

`1 BTC` - `0.909090909 BTC` = `0.09090909 BTC` profit

Reply:
I agree with both your statements: It is hard to follow the amounts and it's hard to understand why it's correct.
Any idea how to make it better? 😬

I tried to come up with better numbers, but I feel the numbers are not the problem but more the model and the way we structure and express things. I could not come up with something better so far though. We already had a hard time with such kind of tests in ItchySats as far as I remember.

Reply:
I don't know... Maybe add the calculation from my comment and add asserts for that? 

Eitherway... if we have to touch this again, we very likely will need to go through this thought process again 🙈

Reply:
Done [here](https://github.com/get10101/10101/pull/994/commits/581150bd641fa65ea2487d896e7edacf9f0451ab).